File: backend/app/core/costbasis.py
# Cost basis math
# Pure engine — no database or broker imports allowed here.
# Takes plain dataclasses, returns results.
#
# Formula: assignment_strike - CSP_premiums - CC_premiums + roll_debits - roll_credits

from dataclasses import dataclass
from decimal import ROUND_HALF_UP, Decimal


@dataclass(frozen=True)
class CspLegInput:
    strike: Decimal
    premium_per_contract: Decimal
    contracts: int


@dataclass(frozen=True)
class CostBasisResult:
    basis_per_share: Decimal
    total_premium_collected: Decimal


def _round4(value: Decimal) -> Decimal:
    return value.quantize(Decimal("0.0001"), rounding=ROUND_HALF_UP)


def calculate_initial_csp_basis(leg: CspLegInput) -> CostBasisResult:
    basis_per_share = _round4(leg.strike - leg.premium_per_contract)
    total_premium_collected = _round4(leg.premium_per_contract * leg.contracts * 100)
    return CostBasisResult(
        basis_per_share=basis_per_share,
        total_premium_collected=total_premium_collected,
    )

File: wheelbase-electron/src/main/core/costbasis.ts
// Cost basis math.
// Pure engine — no database or broker imports allowed here.
//
// Formula: assignment_strike - CSP_premiums - CC_premiums + roll_debits - roll_credits

import Decimal from 'decimal.js'

Decimal.set({ rounding: Decimal.ROUND_HALF_UP })

export interface CspLegInput {
  strike: string
  premiumPerContract: string
  contracts: number
}

export interface CostBasisResult {
  basisPerShare: string
  totalPremiumCollected: string
}

function round4(value: Decimal): Decimal {
  return value.toDecimalPlaces(4)
}

export function calculateInitialCspBasis(leg: CspLegInput): CostBasisResult {
  const strike = new Decimal(leg.strike)
  const premium = new Decimal(leg.premiumPerContract)

  const basisPerShare = round4(strike.minus(premium))
  const totalPremiumCollected = round4(premium.times(leg.contracts).times(100))

  return {
    basisPerShare: basisPerShare.toString(),
    totalPremiumCollected: totalPremiumCollected.toString()
  }
}
